# **Waze Project**
**Course 2 - Get Started with Python**

Bem-vindo ao Projeto Waze!   
Sua equipe de análise de dados do Waze ainda está nos estágios iniciais do projeto de rotatividade de usuários. Anteriormente, você era solicitado a preencher uma proposta de projeto por seu supervisor. Você recebeu um aviso de que sua proposta de projeto foi aprovada e que sua equipe recebeu acesso aos dados de usuário do Waze. Para obter insights claros, os dados do usuário devem ser inspecionados e preparados para o próximo processo de análise exploratória de dados (EDA). Um notebook Python foi preparado para guiá-lo neste projeto. Responda às perguntas e crie um resumo executivo para a equipe de dados do Waze.

# **Course 2 End-of-course project: Inspect and analyze data**

Nesta atividade, você examinará os dados fornecidos e os preparará para análise. Esta atividade ajudará a garantir que as informações estejam:  

1. Prontas para responder a perguntas e gerar insights  
2. Pronto para visualizações  
3. Pronto para futuros testes de hipóteses e métodos estatísticos  
<br/>

**O objetivo** deste projeto é investigar e compreender os dados fornecidos, de forma a usar um dataframe construído em Python, realizar uma inspeção superficial do conjunto de dados fornecido e informar os membros da equipe sobre suas descobertas.
<br/>

*Esta atividade tem três partes:*

**Parte 1:** Entender a situação

**Parte 2:** Entender os dados

**Parte 3:** Entender as variáveis

# **PACE stages**

## **Plan**


### **Task 1. Entender a situação**

*Antes de iniciar a análise compreenda o dicionário dos dados:*

<img src="Dicionario de dados.png" width="593" height="1000" align=left>

## **PACE: Analyze**

### **Task 2a. Importar as bibliotecas e carregar os dados**
Importe as bibliotecas pandas e numpy e carrega o dataset


In [4]:
# Import packages for data manipulation
import pandas as pd 
import numpy as np


In [5]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Informações básicas**
Descubra o shape do dataset, tipo das colunas e se há valores ausentes

In [6]:
df.head(20)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0    0  retained       283     226      296.748273                     2276   
1    1  retained       133     107      326.896596                     1225   
2    2  retained       114      95      135.522926                     2651   
3    3  retained        49      40       67.589221                       15   
4    4  retained        84      68      168.247020                     1562   
5    5  retained       113     103      279.544437                     2637   
6    6  retained         3       2      236.725314                      360   
7    7  retained        39      35      176.072845                     2999   
8    8  retained        57      46      183.532018                      424   
9    9   churned        84      68      244.802115                     2997   
10  10  retained        23      20      117.225772                     1946   
11  11  retained       125     100      357.179031                      487   
12  12  retained        51      41      109.881855                     2760   
13  13  retained        80      64      132.830506                     3154   
14  14   churned        72      66      401.351809                     2524   
15  15  retained        23      22       95.292159                       50   
16  16  retained        30      25       41.538372                     1554   
17  17   churned       272     219      564.271485                       88   
18  18  retained        48      38      259.065753                     2077   
19  19  retained       106      98      170.925470                       87   

    total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                      208                       0       2628.845068   
1                       19                      64      13715.920550   
2                        0                       0       3059.148818   
3                      322                       7        913.591123   
4                      166                       5       3950.202008   
5                        0                       0        901.238699   
6                      185                      18       5249.172828   
7                        0                       0       7892.052468   
8                        0                      26       2651.709764   
9                       72                       0       6043.460295   
10                       0                      36       8554.914440   
11                       3                       7       2284.819767   
12                      64                       0       2512.308716   
13                      39                      16       8531.248070   
14                      54                      24       5922.811202   
15                     193                       0       2188.829771   
16                      61                      19       1213.549279   
17                       0                      90       3529.338980   
18                     131                      11       9140.608725   
19                       0                       6       8566.541401   

    duration_minutes_drives  activity_days  driving_days   device  
0               1985.775061             28            19  Android  
1               3160.472914             13            11   iPhone  
2               1610.735904             14             8  Android  
3                587.196542              7             3   iPhone  
4               1219.555924             27            18  Android  
5                439.101397             15            11   iPhone  
6                726.577205             28            23   iPhone  
7               2466.981741             22            20   iPhone  
8               1594.342984             25            20  Android  
9               2341.838528              7             3   iPhone  
10              5352.567720              9             9   iPhone  
11             

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


#### **Resultados:**
1. 700 linhas ausentes na coluna "label"
2. 14999 linhas e 13 colunas dos tipos inteiro, object e float

### **Task 2c. Valores nulos**

Crie um dataframe somente das linhas ausentes e compare estatisticamente com o restante a fim de descobrir possíveis gaps


In [8]:
# Isolate rows with null values
null_df = df[df['label'].isnull()]

# Display summary stats of rows with null values
null_df.describe()


ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [9]:
# Isolate rows without null values
not_null_df = df[~df['label'].isnull()]

# Display summary stats of rows without null values
not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

#### **Resultados:**
Os valores estatísticos são parecidos, não irá interferir de forma significativa na análise

### **Task 2d. Por dispositivo**

Cheque as populações de 'iphone' e 'android' no dataframe de linhas ausentes, calcule o percentual de cada, compare com o dataset inicial

In [10]:
# Get count of null values by device
print('Iphone ' + str(null_df[null_df['device'] == 'iPhone'].count().values[0]))
print('Iphone ' + str(null_df[null_df['device'] != 'iPhone'].count().values[0]))

Iphone 447
Iphone 253


In [11]:
# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize = True)

iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

In [12]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize = True)


iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

#### **Resultados:**
A porcentagem de valores ausentes por cada dispositivo é consistente com sua representação nos dados em geral. Não há nada que sugira uma causa não aleatória dos dados ausentes.

### **Task 2d. Por label**
Calcule o percentual de retenção x abandono e também a média dos valores para cada um deles


In [15]:
# Calculate counts of churned vs. retained
df['label'].value_counts(normalize = True)



retained    0.822645
churned     0.177355
Name: label, dtype: float64

In [16]:
# Calculate median values of all columns for churned and retained users
df.groupby('label').median(numeric_only = True)


ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

Calcule a mediana de quilômetros por viagem no último mês para usuários retidos e com rotatividade.

In [57]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
df.groupby('label').median()


ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  median_km  \
label                                                                       
churned               1607.183785            8.0           6.0  74.109416   
retained              1458.046141           17.0          14.0  75.014702   

          median_days  km_per_drive  
label                                
churned    697.541999     74.109416  
retained   289.549333     75.014702

Calcule a media de quilômetros por dia dirigido

In [17]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for km per driving day
df.groupby('label').median()

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  \
label                                                            
churned               1607.183785            8.0           6.0   
retained              1458.046141           17.0          14.0   

          km_per_driving_day  
label                         
churned           697.541999  
retained          289.549333

Calcule o número médio de viagens por dia dirigido

In [18]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for drives per driving day
df.groupby('label').median()

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  \
label                                                            
churned               1607.183785            8.0           6.0   
retained              1458.046141           17.0          14.0   

          km_per_driving_day  drives_per_driving_day  
label                                                 
churned           697.541999                 10.0000  
retained          289.549333                  4.0625

Calcule a rotatividade por tipo de dispositivo

In [19]:
# For each label, calculate the number of Android users and iPhone users
df.groupby(['label', 'device']).size()


label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [20]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)


label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

#### **Resultados:**
1. O percentual de usuários por dispositivo comparando o dataset nulo com o geral é bem semelhante. Não há nada que sugira que as linhas ausentes não sejam aleatórias.  
2. O dataset contém cerca de 82% de usuário retidos e 18% de abandonos.  
3. Sugiro tomar cuidado ao realizar cálculos de médias neste dataset, utilizei a mediana devido uma grande variação entre o mínimo e o máximo no describe, diria que quase impossível de ocorrer.  
4. Os usuários que abandonaram o aplicativo dirigiram cerca de 240% a mais que os retidos.  
5. Independente de retidos ou não, os usuários representados neste dataset dirigiram bastante.  
6. A proporção entre usuários de iphone e android é consistente entre os retidos e abandonos.

## **PACE: Construct**

**Note**: O estágio de construção não é válido para este exercício.



## **PACE: Execute**

Crie um resumo estratégico a ser repassado aos seus superiores.